### PR TITLE
impl Step for char (make Range*<char> iterable)

### PR DESCRIPTION
**Pull request by CAD97**
[[irlo thread]](https://internals.rust-lang.org/t/mini-rfc-make-range-char-work/12392?u=cad97) [[godbolt asm example]](https://rust.godbolt.org/z/fdveKo)

Add an implementation of the `Step` trait for `char`, which has the effect of making `RangeInclusive<char>` (and the other range types) iterable.

I've used the surrogate range magic numbers as magic numbers here rather than e.g. a `const SURROGATE_RANGE = 0xD800..0xE000` because these numbers appear to be used as magic numbers elsewhere and there doesn't exist constants for them yet. These files definitely aren't where surrogate range constants should live.

`ExactSizeIterator` is not implemented because `0x10FFFF` is bigger than fits in a `usize == u16`. However, given we already provide some `ExactSizeIterator` that are not correct on 16 bit targets, we might still want to consider providing it for `Range`[`Inclusive`]`<char>`, as it is definitely _very_ convenient. (At the very least, we want to make sure `.count()` doesn't bother iterating the range.)

The second commit in this PR changes a call to `Step::forward` to use `Step::forward_unchecked` in `RangeInclusive::next`. This is because without this patch, iteration over all codepoints (`'\0'..=char::MAX`) does not successfully optimize out the panicking branch. This was mentioned in the PR that updated `Step` to its current design, but was deemed not yet necessary as it did not impact codegen for integral types.

More of `Range*`'s implementations' calls to `Step` methods will probably want to see if they can use the `_unchecked` version as (if) we open up `Step` to being implemented on more types.

---

cc @rust-lang/libs, this is insta-stable and a fairly significant addition to `Range*`'s capabilities; this is the first instance of a noncontinuous domain being iterable with `Range` (or, well, anything other than primitive integers). I don't think this needs a full RFC, but it should definitely get some decent eyes on it.

### CODE REVIEW

**Comment by dtolnay** [comment]
Seems okay to me. Note: I have not reviewed the implementation whatsoever.

**Comment by dtolnay** [comment]
This PR adds `unsafe impl Step for char`, where Step is the unstable trait that's the bound for the implementation of Iterator for various Range types. This makes e.g. `'\0'..=char::MAX` a usable iterator.

@rfcbot fcp merge

**Comment by SimonSapin** [comment]
Is `\uE000` the successor of `\D7FF`? That seems worth a doc-comment on the `impl` with an explanation. Or perhaps on the `Range*` structs, for more visibility.

We’ve been bitten before by inference regressions or name resolution regressions when adding an impl where both the trait and type have existed for a long time, so a Crater run could be good here.

**Comment by CAD97** [comment]
> Is `\u{E000}` the successor of `\u{D7FF}`?

Yes, and this is pretty intrinsic to `char`; as called out [in the docs](https://doc.rust-lang.org/std/primitive.char.html), `char` is a [Unicode Scalar Value](https://www.unicode.org/glossary/#unicode_scalar_value), which is

> Any Unicode [code](https://www.unicode.org/glossary/#code_point) point except high-surrogate and low-surrogate code points. In other words, the ranges of integers 0 to D7FF<sub>16</sub> and E000<sub>16</sub> to 10FFFF<sub>16</sub> inclusive. (See definition D76 in [Section 3.9, Unicode Encoding Forms](http://www.unicode.org/versions/latest/ch03.pdf#G7404).)

> That seems worth a doc-comment

It would be nice to further call this out, sure, but I definitely don't know where best to do so.

> a Crater run could be good here

No arguments here, though I cannot think of how this could potentially lead to any inference failures.

**Comment by SimonSapin** [comment]
I understand very well that surrogate code points are excluded from `char` a.k.a. Unicode scalar values. It’s the successor relationship that wasn’t obvious to me. It makes sense, but some other semantics could also not be absurd. For example have `\uE000` be the 2049th successor of `\uD7FF`, and make everything in-between "invalid" similar to how `\u10FFFF` doesn’t have a successor.

Anyway, these semantics sound good :+1:

**Comment by CAD97** [comment]
The definition of the successor/predecessor operations used by `Step` is that *successor* returns a value that compares greater than or equal, and *predecessor* returns a value that compares less than or equal.

Furthermore, one of the rules is that if there exist some `a, b` where `b > a`, then `Step::forward(a, 1).is_some()` and `Step::forward_unchecked(a, 1)` is safe. I believe that this makes is so that successor/predecessor must walk the full way from MIN to MAX without exposing a hole in the center somewhere.

Certainly, this makes sense for `Range`'s (fused) iterator semantics.

File: src/libcore/iter/range.rs [hunk 1]
```diff
@@ -400,6 +401,73 @@ step_integer_impls! {
...
+    fn steps_between(&start: &char, &end: &char) -> Option<usize> {
+        let start = start as u32;
+        let end = end as u32;
+        if start <= end {
+            let count = end - start + 1;
+            if start < 0xD800 && 0xE000 <= end {
+                usize::try_from(count - 0x800).ok()
+            } else {
+                usize::try_from(count).ok()
+            }
```
Comment (dtolnay) [inline src/libcore/iter/range.rs:416]: This looks wrong according to the spec for Step::steps_between.

https://github.com/rust-lang/rust/blob/45127211566c53bac386b66909a830649182ab7a/src/libcore/iter/range.rs#L31-L32

The code here does not return `Some(0)` if start == end.

**Review by dtolnay — CHANGES_REQUESTED** [review]

**Comment by dtolnay** [comment]
Rest of code looks fine but :boom: there is a lot of `unsafe` code that could use more eyes.

File: src/libcore/iter/range.rs (hunk 1 above)
Comment (CAD97) [reply src/libcore/iter/range.rs:416]: Good catch!
I was too focused on the inclusive case, I forgot `steps_between` is not the inclusive range length 😅

**Review by dtolnay — APPROVED** [review]

**Comment by CAD97** [comment]
(I really need to figure out why my format-on-save is disagreeing with `./x.py fmt`...)

File: src/libcore/tests/iter.rs [hunk 2]
```diff
@@ -1932,6 +1932,18 @@ fn test_range() {
     );
 }
 
+#[test]
+fn test_char_range() {
+    use std::char;
+    assert!(('\0'..=char::MAX).eq((0..=char::MAX as u32).filter_map(char::from_u32)));
+    assert!(('\0'..=char::MAX).rev().eq((0..=char::MAX as u32).filter_map(char::from_u32).rev()));
```
Comment (RalfJung) [inline src/libcore/tests/iter.rs:1939]: Woah doesn't this loop over all `char`?^^

Comment (RalfJung) [reply src/libcore/tests/iter.rs:1939]: Cc https://github.com/rust-lang/rust/pull/72812

**Pull request merged**